core/src/main/scala/com/avsystem/commons/SharedExtensions.scala

```diff
@@ -461,6 +461,19 @@ object SharedExtensionsUtils extends SharedExtensions {
       */
     def toOptArg: OptArg[A] =
       if (tr.isFailure) OptArg.Empty else OptArg(tr.get)
+
+    /**
+      * Apply side-effect only if Try is a failure.
+      *
+      * Don't use .failed projection here, because it unnecessarily creates Exception in case of Success,
+      * which is an expensive operation.
+      */
+    def tapFailure(action: Throwable => Unit): Try[A] = tr match {
+      case Success(_) => tr
+      case Failure(throwable) =>
+        action(throwable)
```

Comment:
What do you want to do when `action(throwable)` throws an exception? Not sure if current choice would be my default, but I'd like to hear your thoughts first. Please document your choice as well.

Reply:
My thoughts were that any `.tap` method should never alter the result. The libraries are also on the fence: `Task#tapError` handles, `Flow#wireTap` I think too but `Iterator#tapEach` and `Observable#doOnNext` do not. I'd probably replicate `Task` approach and use the same name, but I'm okay with other behaviour if it's documented and tested.